### PR TITLE
Fix lock file comparisons

### PR DESCRIPTION
**Pull request by panopticoncentral**
## Bug
Fixes: https://github.com/NuGet/Home/issues/6491
Regression: No

## Fix
Details:

There are several issues with comparing the data in lock files causing lock files to appear dirty when they aren't. This causes the lock file to be written when not dirty which causes Intellisense builds in Visual Studio that shouldn't run.

## Testing/Validation
Tests Added: No
Reason for not adding tests: Not sure how best to add tests for this scenario. 
Validation done:  Ran tests, manually verified

### CODE REVIEW

File: src/NuGet.Core/NuGet.Frameworks/AssetTargetFallbackFramework.cs [hunk 1]
```diff
@@ -80,7 +80,7 @@ public FallbackFramework AsFallbackFramework()
 
         public override bool Equals(object obj)
         {
-            return Equals(obj as FallbackFramework);
+            return Equals(obj as AssetTargetFallbackFramework);
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Frameworks/AssetTargetFallbackFramework.cs:83]: 👍

File: src/NuGet.Core/NuGet.ProjectModel/ProjectRestoreMetadata.cs [hunk 2]
```diff
@@ -161,7 +161,7 @@ public bool Equals(ProjectRestoreMetadata other)
                    OutputPath == other.OutputPath &&
                    ProjectName == other.ProjectName &&
                    ProjectUniqueName == other.ProjectUniqueName &&
-                   EqualityUtility.SequenceEqualWithNullCheck(Sources, other.Sources) &&
```
Comment (nkolev92) [inline src/NuGet.Core/NuGet.ProjectModel/ProjectRestoreMetadata.cs]: Why is this needed?

**Review by emgarten — APPROVED** [review]
🏆 Great catch on these @panopticoncentral, I really appreciate the fix! 🍰

**Review by nkolev92 — APPROVED** [review]
LGTM. 

I think we'll need to do a pass to ensure that there's no more low hanging fruits in this area.

**Comment by nkolev92** [comment]
@panopticoncentral 

One thing I'm curious about is that "unneeded" restores shouldn't be happening at all. 

We do an early no-op check that should catch the restores that would end producing the same result. 
That'd bail running the whole restore and definitely not update the assets file. 
We've made some very recent changes here (just this last week), that fixed an issue there, so you could've been hitting that.  

Can you give us some sort of a repro to investigate this further. 
Basically, we shouldn't even be getting this far.

**Comment by panopticoncentral** [comment]
@nkolev92 The customer scenario is that they have a solution with a large number of projects. They clean one project and then build that one project. That causes a solution restore, perhaps because of the clean? If that seems suspicious, I will try and come up with a repro solution (can't give out the current repro solution).

**Comment by nkolev92** [comment]
Ah, it's because of the clean, yes. 
That explains it. 

If you have a solution handy, that's great, we'll try to spend some time looking into these comparisons again.
